File: internal/web/contacts.go
package web

import (
	"database/sql"
	"encoding/json"
	"errors"
	"net/http"
	"strconv"
	"strings"

	"github.com/slaguardia/scout/internal/store"
)

// handleCompanyContacts is the company's contact list (M51). Dispatched from
// handleCompany for /api/companies/{id}/contacts.
//
//	GET  → list the company's active contacts
//	POST → create a contact {name, role, email}
func (s *Server) handleCompanyContacts(w http.ResponseWriter, r *http.Request, companyID string) {
	switch r.Method {
	case http.MethodGet:
		contacts, err := s.DB.ListContacts(companyID)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		writeJSON(w, http.StatusOK, contacts)
	case http.MethodPost:
		var in store.ContactInput
		if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, 4<<10)).Decode(&in); err != nil {
			http.Error(w, "invalid JSON: "+err.Error(), http.StatusBadRequest)
			return
		}
		c, err := s.DB.CreateContact(companyID, in)
		if err != nil {
			writeContactErr(w, err)
			return
		}
		writeJSON(w, http.StatusOK, c)
	default:
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
	}
}

// handleContact edits or archives one contact. PUT/DELETE /api/contacts/{id}.
func (s *Server) handleContact(w http.ResponseWriter, r *http.Request) {
	id := strings.Trim(strings.TrimPrefix(r.URL.Path, "/api/contacts/"), "/")
	if id == "" || strings.Contains(id, "/") {
		http.NotFound(w, r)
		return
	}
	switch r.Method {
	case http.MethodPut:
		var in store.ContactInput
		if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, 4<<10)).Decode(&in); err != nil {
			http.Error(w, "invalid JSON: "+err.Error(), http.StatusBadRequest)
			return
		}
		c, err := s.DB.UpdateContact(id, in)
		if err != nil {
			writeContactErr(w, err)
			return
		}
		writeJSON(w, http.StatusOK, c)
	case http.MethodDelete:
		if err := s.DB.ArchiveContact(id); err != nil {
			if errors.Is(err, sql.ErrNoRows) {
				http.NotFound(w, r)
				return
			}
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		writeJSON(w, http.StatusOK, map[string]string{"id": id})
	default:
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
	}
}

// writeContactErr maps store contact errors to HTTP codes: unknown id → 404,
// duplicate email → 409, validation → 400, else 500.
func writeContactErr(w http.ResponseWriter, err error) {
	switch {
	case errors.Is(err, sql.ErrNoRows):
		http.Error(w, "not found", http.StatusNotFound)
	case errors.Is(err, store.ErrDuplicateContact):
		http.Error(w, err.Error(), http.StatusConflict)
	case strings.Contains(err.Error(), "contact "):
		http.Error(w, err.Error(), http.StatusBadRequest)
	default:
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
}

// handlePostingOutreachLog is a posting's per-contact send log (M51).
// Dispatched from handlePosting for /api/postings/{id}/outreach-log.
//
//	GET  → list the posting's sends (newest first)
//	POST → log a send {contact_id, sent_at?, note?, followup_due_at?, no_followup?}
func (s *Server) handlePostingOutreachLog(w http.ResponseWriter, r *http.Request, postingID string) {
	switch r.Method {
	case http.MethodGet:
		entries, err := s.DB.ListOutreachForPosting(postingID)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		writeJSON(w, http.StatusOK, entries)
	case http.MethodPost:
		var body struct {
			ContactID string `json:"contact_id"`
			store.OutreachInput
		}
		if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, 8<<10)).Decode(&body); err != nil {
			http.Error(w, "invalid JSON: "+err.Error(), http.StatusBadRequest)
			return
		}
		if strings.TrimSpace(body.ContactID) == "" {
			http.Error(w, "contact_id is required", http.StatusBadRequest)
			return
		}
		e, err := s.DB.LogOutreach(postingID, body.ContactID, body.OutreachInput)
		if err != nil {
			writeOutreachErr(w, err)
			return
		}
		writeJSON(w, http.StatusOK, e)
	default:
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
	}
}

// handleOutreachLog edits or deletes one logged send. PUT/DELETE
// /api/outreach-log/{id}.
func (s *Server) handleOutreachLog(w http.ResponseWriter, r *http.Request) {
	raw := strings.Trim(strings.TrimPrefix(r.URL.Path, "/api/outreach-log/"), "/")
	id, err := strconv.ParseInt(raw, 10, 64)
	if err != nil {
		http.NotFound(w, r)
		return
	}
	switch r.Method {
	case http.MethodPut:
		var e store.OutreachEntryEdit
		if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, 8<<10)).Decode(&e); err != nil {
			http.Error(w, "invalid JSON: "+err.Error(), http.StatusBadRequest)
			return
		}
		out, err := s.DB.UpdateOutreachEntry(id, e)
		if err != nil {
			writeOutreachErr(w, err)
			return
		}
		writeJSON(w, http.StatusOK, out)
	case http.MethodDelete:
		if err := s.DB.DeleteOutreachEntry(id); err != nil {
			if errors.Is(err, sql.ErrNoRows) {
				http.NotFound(w, r)
				return
			}
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		writeJSON(w, http.StatusOK, map[string]int64{"id": id})
	default:
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
	}
}

// writeOutreachErr maps store outreach errors: unknown id → 404, the
// contact/date validation messages → 400, else 500.
func writeOutreachErr(w http.ResponseWriter, err error) {
	switch {
	case errors.Is(err, sql.ErrNoRows):
		http.Error(w, "not found", http.StatusNotFound)
	case strings.HasSuffix(err.Error(), "must be a YYYY-MM-DD date"),
		strings.HasPrefix(err.Error(), "contact not found"):
		http.Error(w, err.Error(), http.StatusBadRequest)
	default:
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
}

// handleFollowupInterval gets or sets the default follow-up interval in business
// days (M51). GET → {days}; PUT {days} → stores it (0–90; 0 disables auto-arm).
func (s *Server) handleFollowupInterval(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case http.MethodGet:
		n, err := s.DB.FollowupIntervalDays()
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		writeJSON(w, http.StatusOK, map[string]int{"days": n})
	case http.MethodPut:
		var body struct {
			Days int `json:"days"`
		}
		if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, 1<<10)).Decode(&body); err != nil {
			http.Error(w, "invalid JSON: "+err.Error(), http.StatusBadRequest)
			return
		}
		if err := s.DB.SetFollowupIntervalDays(body.Days); err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}
		writeJSON(w, http.StatusOK, map[string]int{"days": body.Days})
	default:
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
	}
}

File: scout/web/routes/contacts.py
"""Company contacts + per-contact outreach log + the follow-up interval (M51).

Faithful port of internal/web/contacts.go. Go's handleCompany/handlePosting
prefix dispatch becomes explicit FastAPI routes (the contact/outreach-log paths
are deeper than core's company/posting routes, so they coexist — FastAPI matches
the most specific). Int-id paths parse the segment by hand so a non-numeric id is
a 404 (Go's strconv.ParseInt → http.NotFound), not FastAPI's default 422.
"""
from __future__ import annotations

from fastapi import APIRouter, Depends
from starlette.responses import Response

from scout.store import contacts

from ..deps import get_db
from ..responses import json_error, json_response
from .core import _s, decode_json, raw_body

router = APIRouter()


def _parse_int_id(raw_id: str) -> int | None:
    """The int64 id from a path segment, or None when it isn't a clean integer."""
    try:
        return int(raw_id)
    except (TypeError, ValueError):
        return None


# --- company contacts: /api/companies/{id}/contacts --------------------------


@router.get("/api/companies/{company_id}/contacts")
def list_company_contacts(company_id: str, con=Depends(get_db)) -> Response:
    """A company's active contacts (M51)."""
    return json_response(contacts.list_contacts(con, company_id))


@router.post("/api/companies/{company_id}/contacts")
def create_company_contact(
    company_id: str, raw: bytes = Depends(raw_body), con=Depends(get_db)
) -> Response:
    """Create a contact {name, role, email}. 404 unknown company, 409 duplicate
    email, 400 validation (the store's "contact …" ValueError → global 400)."""
    body = decode_json(raw)
    try:
        c = contacts.create_contact(
            con, company_id,
            contacts.ContactInput(name=_s(body, "name"), role=_s(body, "role"), email=_s(body, "email")),
        )
    except contacts.DuplicateContact as e:
        return json_error(str(e), 409)
    return json_response(c)


# --- one contact: /api/contacts/{id} -----------------------------------------


@router.put("/api/contacts/{contact_id}")
def update_contact(contact_id: str, raw: bytes = Depends(raw_body), con=Depends(get_db)) -> Response:
    body = decode_json(raw)
    try:
        c = contacts.update_contact(
            con, contact_id,
            contacts.ContactInput(name=_s(body, "name"), role=_s(body, "role"), email=_s(body, "email")),
        )
    except contacts.DuplicateContact as e:
        return json_error(str(e), 409)
    return json_response(c)


@router.delete("/api/contacts/{contact_id}")
def archive_contact(contact_id: str, con=Depends(get_db)) -> Response:
    """Soft-delete a contact. 404 for an unknown/already-archived id."""
    contacts.archive_contact(con, contact_id)
    return json_response({"id": contact_id})


# --- per-posting outreach log: /api/postings/{id}/outreach-log ---------------


@router.get("/api/postings/{posting_id}/outreach-log")
def list_outreach_log(posting_id: str, con=Depends(get_db)) -> Response:
    """A posting's per-contact send log, newest first."""
    return json_response(contacts.list_outreach_for_posting(con, posting_id))


@router.post("/api/postings/{posting_id}/outreach-log")
def log_outreach(posting_id: str, raw: bytes = Depends(raw_body), con=Depends(get_db)) -> Response:
    """Log a send {contact_id, sent_at?, body?, note?, followup_due_at?, no_followup?}.
    contact_id is required (400); the contact/date validation maps to 400."""
    body = decode_json(raw)
    contact_id = _s(body, "contact_id")
    if contact_id.strip() == "":
        return json_error("contact_id is required", 400)
    e = contacts.log_outreach(
        con, posting_id, contact_id,
        contacts.OutreachInput(
            sent_at=_s(body, "sent_at"), body=_s(body, "body"), note=_s(body, "note"),
            followup_due_at=_s(body, "followup_due_at"), no_followup=bool(body.get("no_followup")),
        ),
    )
    return json_response(e)


# --- one logged send: /api/outreach-log/{id} ---------------------------------


@router.put("/api/outreach-log/{raw_id}")
def update_outreach_entry(raw_id: str, raw: bytes = Depends(raw_body), con=Depends(get_db)) -> Response:
    id = _parse_int_id(raw_id)
    if id is None:
        return json_error("not found", 404)
    body = decode_json(raw)
    out = contacts.update_outreach_entry(
        con, id,
        contacts.OutreachEntryEdit(
            sent_at=_s(body, "sent_at"), body=_s(body, "body"), note=_s(body, "note"),
            followup_due_at=_s(body, "followup_due_at"), done=bool(body.get("done")),
        ),
    )
    return json_response(out)


@router.delete("/api/outreach-log/{raw_id}")
def delete_outreach_entry(raw_id: str, con=Depends(get_db)) -> Response:
    id = _parse_int_id(raw_id)
    if id is None:
        return json_error("not found", 404)
    contacts.delete_outreach_entry(con, id)
    return json_response({"id": id})


# --- follow-up interval: /api/followup-interval ------------------------------


@router.get("/api/followup-interval")
def get_followup_interval(con=Depends(get_db)) -> Response:
    return json_response({"days": contacts.followup_interval_days(con)})


@router.put("/api/followup-interval")
def set_followup_interval(raw: bytes = Depends(raw_body), con=Depends(get_db)) -> Response:
    """Store the default follow-up interval in business days (0–90; 0 disables
    auto-arm). Out-of-range → 400 (the store ValueError → global 400)."""
    body = decode_json(raw)
    days = body.get("days", 0)
    if isinstance(days, bool) or not isinstance(days, int):
        return json_error("invalid JSON: days must be an integer", 400)
    contacts.set_followup_interval_days(con, days)
    return json_response({"days": days})
